common/src/main/scala/explore/components/AgGridHelpers.scala

```diff
@@ -0,0 +1,258 @@
+
+  private def newId: NonEmptyString = NonEmptyString
+    .from("Id" + java.util.UUID.randomUUID().toString().replace("-", ""))
+    .getOrElse("idddd")
+
+  // TODO: Need to make a Backend for this to expose the ag-grid lifecycle methods
+  // (if that's how we expose them). Also probably want to edit State instead of view
+  // directly, as with the FormInputEV version.
+  private def makeFormInputEnumEditor[A](exclude: Set[A], includeCurrent: Boolean)(implicit
+    enum:                                         Enumerated[A],
+    display:                                      Display[A]
+  ) = {
+    val excludeFn: A => Set[A] = a => if (includeCurrent) exclude - a else exclude
+    ScalaComponent
+      .builder[View[A]]
+      .render_P(view => EnumViewSelect(id = newId, value = view, exclude = excludeFn(view.get)))
+      .build
+      .cmapCtorProps[ICellEditorParams](parms => parms.value.asInstanceOf[View[A]])
+      .toJsComponent
+      .raw
+  }
+
+  private def makeFormInputEVEditor[A](
+    validator:     ValidFormatInput[A],
+    changeAuditor: ChangeAuditor[A]
+  )(implicit eq:   Eq[A]) = {
+
+    @Lenses
+    case class State(original: A, current: A, editing: A)
+
+    class Backend($ : BackendScope[View[A], State])(implicit val eq: Eq[A]) {
+      val id = newId
+
+      // We need to add a little delay before looking for the input element. I think
+      // ag-grid is using some virtualization of it's own, and the input is not in
+      // the DOM immediately.
+      def setCursor(): Callback =
+        CallbackTo {
+          val inputOpt = Option(document.querySelector(s"#${id.value}").asInstanceOf[html.Input])
+          inputOpt.map { i =>
+            i.focus();
+            val len = i.value.length
+            i.setSelectionRange(0, len)
+          }
+        }.void.delayMs(50).completeWith(_ => Callback(()))
+
+      // We have to wait until getValue is called before setting the value or
+      // we get "kicked out of" editing due to re-rendering of the table.
+      // So, we need to `createInitialStateFromProps` to create a View
+      // in the State that can be edited by the FormInputEv. We also
+      // keep the original value so that we can compare and return the
+      // appropriate response in the `isCancelAfterEnd` "lifecycle" method.
+
+      def onTextChange(text: String): Callback =
+        validator.getValidated(text) match {
+          case Valid(a)   => $.setStateL(State.current)(a)
+          // If the current text is invalid, we reset to the original to behave more
+          // like the inputs that are not in tables.
+          case Invalid(_) => $.state.flatMap(s => $.setStateL(State.current)(s.original))
+        }
+
+      // ag-grid lifecycle method.
+      // Note that it doesn't matter what `getValue` actually returns, since
+      // 'valueSetter' is a noop. We just need to set the value in the View
+      // in the props.
+      // Is there a way to avoid calling runNow()?
```

Comment:
Is there a way to avoid calling runNow() here and in `isCancelAfterEnd`? Is calling runNow() safe in this instance?

Reply:
Very often when interacting with the js world you need to call `runNow()`, basically if this is to be sent to `ag-grid` you probably need to call `runNov` to execute the callback

Reply:
Thanks. I thought that was probably the case, and that it was being sent outside of react it was safe to call. But, I thought I should make sure.